ntoskrnl/config/ntapi.c

```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+                {
+                    ObDereferenceObject(KeyObject);
+                    Status = STATUS_INSUFFICIENT_RESOURCES;
+                    _SEH2_YIELD(goto Quit);
+                }
+
+                SubNames = ExAllocatePool(NonPagedPool, Count * sizeof(UNICODE_STRING));
+                if (!SubNames)
+                {
+                    ObDereferenceObject(KeyObject);
+                    ExFreePool(LocalSubObjects);
+                    Status = STATUS_INSUFFICIENT_RESOURCES;
+                    _SEH2_YIELD(goto Quit);
+                }
+
+                for (int i = 0; i < Count; i++)
+                {
+                    Status = ProbeAndCaptureObjectAttributes(
+                        &LocalSubObjects[i], &SubNames[i], PreviousMode, &SlaveObjects[i], FALSE);
+                    if (!NT_SUCCESS(Status))
+                        break;
+                }
+            }
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            /* Return the exception code */
+            Status = _SEH2_GetExceptionCode();
+        }
+        _SEH2_END;
+
+        if (!NT_SUCCESS(Status))
+        {
+            if (KeyObject)
+                ObDereferenceObject(KeyObject);
+            if (LocalSubObjects)
+                ExFreePool(LocalSubObjects);
+            if (SubNames)
+                ExFreePool(SubNames);
+            goto Quit;
+        }
+    }
+
+    /* Lock the registry while we're setting up notifications */
+    CmpLockRegistryExclusive();
+
+    /* Lock the Kcb while we are initializing/updating NotifyBlock */
+    CmpAcquireKcbLockExclusive(KeyObject->KeyControlBlock);
+
+    /* Open subordinate objects */
+    if (Count > 0)
+    {
+        LocalSubObjectsKeyBody = ExAllocatePool(NonPagedPool, Count * sizeof(PCM_KEY_BODY));
+        if (!LocalSubObjectsKeyBody)
+        {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Failure;
+        }
+        RtlZeroMemory(LocalSubObjectsKeyBody, Count * sizeof(PCM_KEY_BODY));
+
+        for (int i = 0; i < Count; i++)
+        {
+            /* Open a kernel-mode handle, since we're using it internally without returning it to the user */
+            HANDLE SubHandle;
+            LocalSubObjects[i].Attributes |= OBJ_KERNEL_HANDLE;
+            Status = ObOpenObjectByName(
+                &LocalSubObjects[i], CmpKeyObjectType, KernelMode, NULL, KEY_NOTIFY, NULL, &SubHandle);
+            if (!NT_SUCCESS(Status))
+                goto Failure;
+
+            Status = ObReferenceObjectByHandle(
+                SubHandle, KEY_NOTIFY, CmpKeyObjectType, KernelMode, (PVOID *)&LocalSubObjectsKeyBody[i], NULL);
+            /* clsoe the handle as we don't need it anymore */
+            ZwClose(SubHandle);
+            if (!NT_SUCCESS(Status))
+                goto Failure;
+
+            /* Check if this is a duplicate object */
+            if (LocalSubObjectsKeyBody[i]->KeyControlBlock == KeyObject->KeyControlBlock)
```

Comment:
I suppose you are checking here if the KCB of the sub key matches with the primary key object to avoid locking it again, in order to prevent deadlocks? Why not just simply check if the sub KCB is different from the main one, and lock the KCB for exclusive use right away?

Reply:
I guess it would cause a deadlock if a subordinate is pointing the same key as master, but well I added this check because it assumes each provided key is a different key. You're supposed to provide different keys when you want to watch multiple keys. Come to think about this, Windows locks the Hive in this function and also it prevents you from providing a subordinate in the same Hive as the master key, and this should be the reason.

Reply:
Please provide a testcase in your KMTEST that indeed Windows throws `STATUS_INVALID_PARAMETER` if the caller submits the same subordinate key as that of the master one.

>I guess it would cause a deadlock

Of course it will cause a deadlock, as KCBs are protected by using a push lock and if those aren't used judiciously then you're gonna freeze the whole system as both threads are waiting on each other.

Reply:
> Please provide a testcase in your KMTEST that indeed Windows throws STATUS_INVALID_PARAMETER if the caller submits the same subordinate key as that of the master one, if you haven't done it so.

I didn't, I'll do OK. Also Windows returns error even if the subordinate key is in the same Key Hive as the master, so it doesn't have to be the same key. The error code is different on Windows 2000 and Windows 11, where Windows 2000 returns `STATUS_INVALID_PARAMETER` but Windows 11 returns `STATUS_INVALID_OBJECT_NAME`.

Reply:
The test is provided now but in NTDLL_APITEST instead of KMTEST. I'll mark this as resolved if there's no other feedback.